File: Python/Algorithms/1323.py
class Solution:
    def maximum69Number (_, num: int) -> int:
        if num == 6 or num == 9:
            return 9
        if num == 99 or num == 999 or num == 9999:
            return num
        divmod_ = divmod
        # Get correct multiplier for 9. (i.e how manhy digits is num)
        d = 10 ** (num.bit_length() // 4)
        d = d * 9
        div, t = divmod_(num, d)
        # Continuously reduce d and t until we find the part
        # of num we can't divide with d.
        while div:
            d, _ = divmod_(d, 10)
            div, t = divmod_(t, d)
        # Then, we just add d // 3 to the final number.
        # I.e num = 9969 => d is reduced to 90
        # so add 90 // 3 (30)
        return num + (d // 3)

File: Rust/Algorithms/1323.rs
impl Solution {
    pub fn maximum69_number (num: i32) -> i32 {
        let (mut digits, mut temp) = (0, num);
        // Count number of digits (+ 1)
        while temp != 0 {
            digits += 1;
            temp /= 10;
        }
        // Find appropriate 10-multiple of 9.
        let mut d: i32 = 9 * (10 as i32).pow(digits-1);
        let (mut div, mut t) = (num / d, num % d);

        // Calculate number.
        while div == 1 {
            d /= 10;
            if d == 0 {
                return num;
            }
            div = t / d;
            t %= d;
        }
        num + (d / 3)
    }
}
